src/repr/type.ml

```diff
@@ -371,3 +371,60 @@ module Unboxed = struct
 
   let size_of = Type_size.unboxed
 end
+
+let ref : type a. a t -> a ref t = fun a -> map a ref (fun t -> !t)
+let lazy_t : type a. a t -> a Lazy.t t = fun a -> map a Lazy.from_val Lazy.force
+
+let seq : type a. a t -> a Seq.t t =
+ fun a ->
+  let elt_equal = unstage @@ equal a in
+  let elt_compare = unstage @@ compare a in
+  let rec compare (s1 : a Seq.t) (s2 : a Seq.t) =
+    match (s1 (), s2 ()) with
+    | Nil, Nil -> 0
+    | Cons _, Nil -> 1
+    | Nil, Cons _ -> -1
+    | Cons (x, xf), Cons (y, yf) ->
+        let ord = elt_compare x y in
+        if ord <> 0 then ord else compare xf yf
+  in
+  let rec equal (s1 : a Seq.t) (s2 : a Seq.t) =
+    match (s1 (), s2 ()) with
+    | Nil, Nil -> true
+    | Cons _, Nil | Nil, Cons _ -> false
+    | Cons (x, xf), Cons (y, yf) -> elt_equal x y && equal xf yf
+  in
+  map ~compare:(stage compare) ~equal:(stage equal) (list a) List.to_seq
```

Comment:
You could use key/value mapping for the hashtbl for the json ser/deser. 

Maybe on a second pass, as you said.

Reply:
Indeed, I think there are quite a few such local improvements that could be made, but would like to avoid getting bogged down in that for now.